charts/kong/templates/_helpers.tpl

```diff
@@ -535,7 +535,22 @@ The name of the service used for the ingress controller's validation webhook
   {{- toYaml .Values.resources | nindent 4 }}
 {{- end -}}
 
+{{/*
+kong.controller2xplus returns "true" if the controller image tag major version is >= 2
+Note that this is a string, not a boolean, because templates vov
+*/}}
+{{- define "kong.controller2xplus" -}}
+{{- $version := "" -}}
+{{- if .Values.ingressController.image.effectiveSemver -}}
+  {{- $version = semver .Values.ingressController.image.effectiveSemver -}}
+{{- else -}}
+  {{- $version = semver .Values.ingressController.image.tag -}}
+{{- end -}}
+{{- ge $version.Major 2 -}}
```

Comment:
could we make this template parametric in the following way, for reusability:
 - accept the major version as parameter
 - accept the `image` object as a parameter

so that it's reusable if we, say, need to check elsewhere if the gateway has major version >= 3?

Reply:
Looked at this briefly and I think I'd like to defer it for now. This function is intentionally a bit weird because we want this to work for the beta, and for that the major version comparison is necessary over a more conventional `semver.Compare`: `2.0.0-beta.1` is not greater than `2.0.0` (and IIRC it's not even greater than `1.3`, ignoring that using that would cause issues if we eventually released 1.4). Realistically our more immediate need, if any, will be to check if the controller version is >= 2.1, not >= 3.

We could work around that by converting to semver first, extracting the major, and passing that in, but IMO more work than its worth at this point--makes more sense to release a simpler version for now and then refactor into something more reusable after 2.0 GA.

Making a "function" in template land is a bit cumbersome, since parameters aren't really a thing--you have only a single parameter, so you have to construct a dict on the fly and don't have the benefit of signature checking. It's clunky but doable. 

Whether we truly want to do that is an open question--my general experience with developing entirely in templates is that it's an exercise in masochism. This is a good example, where we have odd workarounds for not having proper functions and types and lack access to the full feature set of the underlying semver library. I think a reasonable next major project is to begin porting chart functionality into a full Go operator that handles complex decisions in Go and only uses templates for a final "fill in the exact given values" step to create a manifest.